pkgs/applications/office/paperless/default.nix

```diff
@@ -0,0 +1,222 @@
+        --set PATH ${extraBin} --add-flags $out/share/paperless/manage.py
+
+      # A shell snippet that can be sourced to setup a paperless env
+      cat > $out/share/paperless/setup-env.sh <<EOF
+      export PATH=${pythonEnv}/bin:${extraBin}''${PATH:+:}$PATH
+      export paperlessSrc=$out/share/paperless
+      EOF
+    '';
+
+    checkPhase = ''
+      source $out/share/paperless/setup-env.sh
+      tmpDir=$(realpath testsTmp)
+      mkdir $tmpDir
+      export HOME=$tmpDir
+      export PAPERLESS_MEDIADIR=$tmpDir
+      cd $paperlessSrc
+      # Prevent tests from writing to the derivation output
+      chmod -R -w $out
+      # Disable cache to silence a pytest warning ("could not create cache")
+      pytest -p no:cacheprovider
+    '';
+
+    passthru = {
+      inherit withConfig;
+    };
+
+    meta = with lib; {
+      description = "Scan, index, and archive all of your paper documents";
+      homepage = https://github.com/danielquinn/paperless;
+      license = licenses.gpl3;
+      maintainers = [ maintainers.earvstedt ];
+    };
+  };
+
+  # Use Python 3.6 because 3.7 is currently incompatible with these packages:
+  #  - dateparser:
+  #      Although version 0.7.0 in Nixpkgs is fundamentally compatible with 3.7,
+  #      3 out of 23161 tests are failing with 3.7.
+  #  - ruamel_yaml: compatible version unavailable in Nixpkgs
+  #  - factory_boy: compatible version available in repo, but not released yet
+  python = python36.override {
+    # Paperless only supports Django 2.0
+    packageOverrides = self: super: { django = super.django_2_0; };
+  };
+
+  pyocrWithUserTesseract =
+    let
+      pyocr = python.pkgs.pyocr.override { inherit tesseract; };
+    in
+      if pyocr.outPath == python.pkgs.pyocr.outPath then
+        pyocr
+      else
+        # The user has provided a custom tesseract derivation that might be
+        # missing some languages that are required for PyOCR's tests. Disable them to
+        # avoid build errors.
+        pyocr.overrideAttrs (attrs: {
+          doInstallCheck = false;
+        });
+
+  withConfig = { config ? {}, dataDir ? null, paperlessPkg ? paperless, extraCmds ? "" }:
+    let
+      dir = toString dataDir;
+
+      envVars = (optionalAttrs (dataDir != null) {
+        PAPERLESS_CONSUMPTION_DIR = "${dir}/consume";
+        PAPERLESS_MEDIADIR = "${dir}/media";
+        PAPERLESS_STATICDIR = "${dir}/static";
+        PAPERLESS_DBDIR = "${dir}";
+      }) // config;
+
+      envVarDefs = mapAttrsToList (n: v: ''export ${n}="${toString v}"'') envVars;
+      setupEnvVars = builtins.concatStringsSep "\n" envVarDefs;
+
+      setupEnv = ''
+        source ${paperlessPkg}/share/paperless/setup-env.sh
+        ${setupEnvVars}
+        ${optionalString (dataDir != null) ''
+          mkdir -p "$PAPERLESS_CONSUMPTION_DIR" \
+                   "$PAPERLESS_MEDIADIR" \
+                   "$PAPERLESS_STATICDIR" \
+                   "$PAPERLESS_DBDIR"
```

Comment:
This means the directories are created every time we run paperless? Wouldn't it be better to move all this stuff (everything from `withConfig`) into the service? Setting something up should not be done by the package if it's not what upstream does.

Reply:
Done.